File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function buildTree(preorder: number[], inorder: number[]): TreeNode | null {
  const N = preorder.length,
        map = new Map<number, number>();
  let idx = 0;
  
  for(let i = 0; i < N; i++)
    map.set(inorder[i], i)
  
  return dfs(0, N - 1);
  
  function dfs(left: number, right: number): TreeNode | null {
    if(left > right)
      return null;

    const rootVal = preorder[idx++],
        i = map.get(rootVal);
    
    return new TreeNode(rootVal, dfs(left, i - 1), dfs(i + 1, right));
  }
};

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

	int pIdx;
	int iIdx;

	public TreeNode buildTree(int[] preorder, int[] inorder) {
		this.pIdx = 0;
        this.iIdx = 0;
        
		return this.buildTree(preorder, inorder, Integer.MIN_VALUE);
	}

	public TreeNode buildTree(int[] preorder, int[] inorder, int stop) {
        if(this.pIdx >= preorder.length){
            return null;
        }
        
        if(inorder[this.iIdx] == stop) {
            this.iIdx += 1;
            return null;
        }

		int rootVal = preorder[this.pIdx++];
		TreeNode root = new TreeNode(rootVal);

		root.left = this.buildTree(preorder, inorder, rootVal);
		root.right = this.buildTree(preorder, inorder, stop);

		return root;
	}
}